File: tools/http/test-client/js/test/contract.test.js
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

"use strict";

const assert = require("node:assert/strict");
const { describe, it } = require("node:test");

const {
  ACTIONS_VARIABLE,
  ACTION_VARIABLE,
  exchangeFor,
  exchanges,
  renderResponseBody,
  requests,
  scenarioRequest,
} = require("../src");
const { ACTIONS, ACTIONS_JSON } = require("./actions");

process.env[ACTIONS_VARIABLE] = ACTIONS_JSON;

describe("the contract", () => {
  it("decodes the complete runner action table", () => {
    assert.ok(exchanges().length > 0);
  });

  it("does not measure readiness", () => {
    assert.ok(exchanges().some((exchange) => exchange.readiness));
    assert.ok(!requests().some((exchange) => exchange.readiness));
    assert.equal(requests().length, exchanges().length - 1);
  });

  it("selects each singular client action", () => {
    requests().forEach((exchange, index) => {
      assert.deepEqual(
        scenarioRequest(JSON.stringify(ACTIONS[index + 1])),
        exchange,
      );
    });
  });

  it("says when the selected action is not set", () => {
    const previous = process.env[ACTION_VARIABLE];
    delete process.env[ACTION_VARIABLE];
    try {
      assert.throws(
        () => scenarioRequest(),
        new RegExp(`${ACTION_VARIABLE} is not set`),
      );
    } finally {
      if (previous === undefined) {
        delete process.env[ACTION_VARIABLE];
      } else {
        process.env[ACTION_VARIABLE] = previous;
      }
    }
  });

  it("rejects malformed JSON and unknown action fields", () => {
    assert.throws(() => scenarioRequest("{"), /malformed JSON/);
    assert.throws(() => exchanges("{"), /malformed JSON/);
    assert.throws(
      () => scenarioRequest('{"request":{},"response":{},"extra":true}'),
      /unknown field/,
    );
  });

  it("says when the complete action table is not set", () => {
    const previous = process.env[ACTIONS_VARIABLE];
    delete process.env[ACTIONS_VARIABLE];
    try {
      assert.throws(
        () => exchanges(),
        new RegExp(`${ACTIONS_VARIABLE} is not set`),
      );
    } finally {
      process.env[ACTIONS_VARIABLE] = previous;
    }
  });

  // A field this reader names but the contract does not is `undefined` here
  // rather than an error, so only a check on the value catches a rename.
  it("carries every field the contract names", () => {
    for (const exchange of exchanges()) {
      assert.equal(typeof exchange.description, "string");
      assert.ok(exchange.description.trim().length > 0);
      assert.equal(typeof exchange.method, "string");
      assert.equal(typeof exchange.path, "string");
      assert.equal(typeof exchange.status, "number");
      assert.equal(typeof exchange.responseBody, "string");
    }
  });

  it("answers the same exchange with or without a query string", () => {
    const plain = exchangeFor("GET", "/users/123");
    const withQuery = exchangeFor("GET", "/users/123?fields=name&verbose=true");

    assert.equal(plain.status, withQuery.status);
    assert.equal(plain.responseBody, withQuery.responseBody);
  });

  it("looks up the method as well as the path", () => {
    assert.equal(exchangeFor("DELETE", "/items"), null);
  });

  it("describes no exchange for an unknown path", () => {
    assert.equal(exchangeFor("GET", "/nope"), null);
  });

  it("echoes the body that arrived", () => {
    const items = exchangeFor("POST", "/items");

    assert.equal(
      renderResponseBody(items, '{"name": "widget"}'),
      '{"created": true, "payload": {"name": "widget"}}',
    );
  });

  it("echoes an empty object for a request that carried no body", () => {
    const items = exchangeFor("POST", "/items");

    assert.equal(
      renderResponseBody(items, null),
      renderResponseBody(items, ""),
    );
    assert.equal(
      renderResponseBody(items, null),
      '{"created": true, "payload": {}}',
    );
  });

  // `$&` and its siblings are substitution patterns to a JavaScript string
  // replacement, and to nothing at all in the readers the other languages use.
  it("echoes a body the way every other language reads it", () => {
    const items = exchangeFor("POST", "/items");

    assert.equal(
      renderResponseBody(items, '{"name": "a$&b$`c$\'d$$e"}'),
      '{"created": true, "payload": {"name": "a$&b$`c$\'d$$e"}}',
    );
  });
});

File: tools/http/test-client/ruby/test/contract_test.rb
# Copyright The OpenTelemetry Authors
# SPDX-License-Identifier: Apache-2.0

require_relative "test_helper"

class ContractTest < Minitest::Test
  def test_reads_every_contract_field
    refute_empty HTTP_CONTRACT.exchanges

    HTTP_CONTRACT.exchanges.each do |exchange|
      refute_empty exchange.description
      refute_empty exchange.method
      refute_empty exchange.path
      assert_kind_of Integer, exchange.status
      assert_kind_of String, exchange.response_body
      assert_includes [true, false], exchange.readiness
      assert(exchange.body.nil? || exchange.body.is_a?(String))
    end
  end

  def test_exchange_data_is_immutable
    assert_predicate HTTP_CONTRACT.exchanges, :frozen?
    assert_predicate HTTP_CONTRACT.requests, :frozen?
    assert_predicate HTTP_CONTRACT.exchanges.first, :frozen?
    assert_predicate HTTP_CONTRACT.exchanges.first.path, :frozen?
  end

  def test_excludes_readiness_from_measured_requests
    assert HTTP_CONTRACT.exchanges.any?(&:readiness)
    refute HTTP_CONTRACT.requests.any?(&:readiness)
    assert_equal HTTP_CONTRACT.exchanges.length - 1, HTTP_CONTRACT.requests.length
  end

  def test_lookup_ignores_query_but_checks_method
    plain = HTTP_CONTRACT.exchange_for("GET", "/users/123")
    queried = HTTP_CONTRACT.exchange_for("GET", "/users/123?fields=name")

    assert_equal plain.status, queried.status
    assert_equal plain.response_body, queried.response_body
    assert_nil HTTP_CONTRACT.exchange_for("DELETE", "/users/123")
    assert_nil HTTP_CONTRACT.exchange_for("GET", "/missing")
  end

  def test_renders_the_request_body
    exchange = HTTP_CONTRACT.exchange_for("POST", "/items")

    assert_equal(
      '{"created": true, "payload": {"name": "widget"}}',
      exchange.render_response_body('{"name": "widget"}')
    )
    assert_equal(
      '{"created": true, "payload": {}}',
      exchange.render_response_body(nil)
    )
  end

  def test_defines_fixed_headers
    assert_equal(
      { "User-Agent" => "otel-http-conformance/1" },
      HTTP_CONTRACT.client_headers(nil)
    )
    assert_equal(
      {
        "User-Agent" => "otel-http-conformance/1",
        "Content-Type" => "application/json"
      },
      HTTP_CONTRACT.client_headers('{"name":"widget"}')
    )
  end
end
